### PR TITLE
cpptrace: Add recipe

### PR DESCRIPTION
**Pull request by jeremy-rifkin**
Specify library name and version:  **cpptrace/0.2**

I'm the author of [cpptrace](https://github.com/jeremy-rifkin/cpptrace), which is a dependency for [libassert](https://github.com/jeremy-rifkin/libassert) v1.2.

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/cpptrace/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    def validate(self):
+        if self.settings.compiler.cppstd:
+            check_min_cppstd(self, self._min_cppstd)
+        check_min_vs(self, 191)
+        if not is_msvc(self):
+            minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
+            if minimum_version and Version(self.settings.compiler.version) < minimum_version:
+                raise ConanInvalidConfiguration(
+                    f"{self.ref} requires C++{self._min_cppstd}, which your compiler does not support."
+                )
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:63]: For C++11 libs we're not enforcing compiler checks (Those were needed in Conan v1 as it did not model any cppstd attribute, but now with Conan 2 we can just set it to 11 and forget about it, the idea is to phase out v1 as time goes on)

Comment (RubenRBS) [reply recipes/cpptrace/all/conanfile.py:63]:
```suggestion
```

File: recipes/cpptrace/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        if is_msvc(self):
+            tc.variables["USE_MSVC_RUNTIME_LIBRARY_DLL"] = not is_msvc_static_runtime(self)
+        tc.generate()
+        tc = CMakeDeps(self)
+        tc.generate()
+        tc = VirtualBuildEnv(self)
+        tc.generate(scope="build")
+
+    def build(self):
+        apply_conandata_patches(self)
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:81]:
```suggestion
```

No need if no patches :)

File: recipes/cpptrace/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        "shared": True,
+        "fPIC": True,
+    }
+
+    @property
+    def _min_cppstd(self):
+        return 11
+
+    def export_sources(self):
+        export_conandata_patches(self)
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:36]:
```suggestion
```

No need if no patches

File: recipes/cpptrace/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            self.cpp_info.system_libs.append("pthread")
+            self.cpp_info.system_libs.append("dl")
+        if self.settings.os == "Windows":
+            self.cpp_info.system_libs.append("dbghelp")
+
+        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.filenames["cmake_find_package"] = "PACKAGE"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "package"
+        self.cpp_info.names["cmake_find_package"] = "PACKAGE"
+        self.cpp_info.names["cmake_find_package_multi"] = "package"
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:115]: You probably want to fill this one out still :P

File: recipes/cpptrace/all/test_package/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.15)
+
+project(test_package CXX)
+
+find_package(cpptrace REQUIRED CONFIG)
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+
+target_link_libraries(${PROJECT_NAME} PRIVATE cpptrace::cpptrace)
```
Comment (RubenRBS) [inline recipes/cpptrace/all/test_package/CMakeLists.txt:9]: Maybe add a `target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_11)` to ensure it works in C++11?

**Review by RubenRBS — COMMENTED** [review]
Hello friend :)

It looks super good, I just have a few comments (Plus the annoying linter error, pelase fix it or it won't stop complaining, sorry it's annoying)

**Comment by jeremy-rifkin** [comment]
Hi @RubenRBS, thank you so much for the quick review! I'll finish ironing things out and incorporate your feedback!

File: recipes/cpptrace/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+from conan.tools.build import check_min_cppstd
+from conan.tools.scm import Version
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+from conan.tools.env import VirtualBuildEnv
+import os
+
+required_conan_version = ">=1.53.0"
+
+class PackageConan(ConanFile):
+    name = "package"
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:14]: Didnt catch it at first: this should be the lib's name (the description field also need filling)

**Review by RubenRBS — CHANGES_REQUESTED** [review]
Opps, oke extra change needes before this can be merged

File: recipes/cpptrace/all/conanfile.py (hunk 6 above)
Comment (jeremy-rifkin) [reply recipes/cpptrace/all/conanfile.py]: Ope thank you? That fixed at least one weird error I was seeing locally. Sorry to bother you with this but when I try building I'm seeing

```-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: Target declared 'cpptrace::cpptrace'
CMake Error at build/gcc-9-x86_64-17-release/generators/cmakedeps_macros.cmake:67 (message):
  Library 'package_lib' not found in package.  If 'package_lib' is a system
  library, declare it with 'cpp_info.system_libs' property
Call Stack (most recent call first):
  build/gcc-9-x86_64-17-release/generators/cpptrace-Target-release.cmake:24 (conan_package_library_targets)
  build/gcc-9-x86_64-17-release/generators/cpptraceTargets.cmake:26 (include)
  build/gcc-9-x86_64-17-release/generators/cpptrace-config.cmake:16 (include)
  CMakeLists.txt:5 (find_package)


-- Configuring incomplete, errors occurred!

*********************************************************
Recipe 'cpptrace/0.2.0 (test package)' cannot build its binary
It is possible that this recipe is not Conan 2.0 ready
If the recipe comes from ConanCenter, report it at https://github.com/conan-io/conan-center-index/issues
If it is your recipe, check if it is updated to 2.0
*********************************************************

ERROR: cpptrace/0.2.0 (test package): Error in build() method, line 21
        cmake.configure()
        ConanException: Error 1 while executing
```

Not quite sure what I'm doing wrong

Comment (jeremy-rifkin) [reply recipes/cpptrace/all/conanfile.py]: Oh, never mind, silly mistake on my end :)

File: recipes/cpptrace/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+from conan.tools.scm import Version
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+from conan.tools.env import VirtualBuildEnv
+import os
+
+required_conan_version = ">=1.53.0"
+
+class PackageConan(ConanFile):
+    name = "cpptrace"
+    description = "short description"
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:15]:
```suggestion
    description = "Cpptrace is a lightweight C++ stacktrace library supporting C++11 and greater on Linux, macOS, and Windows including MinGW and Cygwin environments. The goal: Make stack traces simple for once"
```

**Comment by jeremy-rifkin** [comment]
I think that pipeline failure isn't from the most recent changes. `conan export recipes/cpptrace/all/conanfile.py --version 0.2.0` works for me. I think everything should now be resolved.

**Review by RubenRBS — COMMENTED** [review]
Note that the job queue is currently quite busy so the next compilation results might take a while to show up :)

**Comment by RubenRBS** [comment]
> I think that pipeline failure isn't from the most recent changes. conan export recipes/cpptrace/all/conanfile.py --version 0.2.0 works for me. I think everything should now be resolved.

Yeah sorry about that, as I said the build queue is currently quite busy so some things are a bit delayed at the moment

**Comment by jeremy-rifkin** [comment]
All good thank you so much!

File: recipes/cpptrace/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+from conan.tools.files import apply_conandata_patches, export_conandata_patches, get, copy, rm, rmdir, replace_in_file
+from conan.tools.build import check_min_cppstd
+from conan.tools.scm import Version
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+from conan.tools.env import VirtualBuildEnv
+import os
+
+required_conan_version = ">=1.53.0"
+
+class PackageConan(ConanFile):
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:13]:
```suggestion
class CpptraceConan(ConanFile):
```

File: recipes/cpptrace/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.microsoft import check_min_vs, is_msvc_static_runtime, is_msvc
+from conan.tools.files import apply_conandata_patches, export_conandata_patches, get, copy, rm, rmdir, replace_in_file
+from conan.tools.build import check_min_cppstd
+from conan.tools.scm import Version
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:6]:
```suggestion
```

File: recipes/cpptrace/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.microsoft import check_min_vs, is_msvc_static_runtime, is_msvc
+from conan.tools.files import apply_conandata_patches, export_conandata_patches, get, copy, rm, rmdir, replace_in_file
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:4]:
```suggestion
from conan.tools.files import get, copy, rm, rmdir
```

File: recipes/cpptrace/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
```
Comment (RubenRBS) [inline recipes/cpptrace/all/conanfile.py:2]:
```suggestion
```

**Review by RubenRBS — COMMENTED** [review]
Retriggering the build with this last changes

**Comment by jeremy-rifkin** [comment]
Thank you so much for those fixes!

**Comment by jeremy-rifkin** [comment]
I think this should fix the CI error for shared=false

**Comment by RubenRBS** [comment]
Got a busy few days, I'll try to give you a helping hand on monday :)

**Comment by jeremy-rifkin** [comment]
Thank you so much Rubén, no rush at all :) I'll may keep fiddling with it over the weekend
giphy

**Comment by jeremy-rifkin** [comment]
I think this should be good to go, fingers crossed

**Review by RubenRBS — DISMISSED** [review]
Thanks a lot! :)

**Review by RubenRBS — APPROVED** [review]

**Pull request merged**